### ISSUE TITLE
[PHP8.2] mark class readonly if all props are readonly

### ISSUE THREAD

**Issue by keradus** [PHP-CS-Fixer/PHP-CS-Fixer#6801]
before
```php
class Post
{
    public function __construct(
        public readonly string $title, 
        public readonly Author $author,
        public readonly string $body,
        public readonly DateTime $publishedAt,
    ) {}
}
```

after
```php
readonly class Post
{
    public function __construct(
        public string $title, 
        public Author $author,
        public string $body,
        public DateTime $publishedAt,
    ) {}
}
```

**Comment by mvorisek** [comment]
I do not think this is a good idea. Especially as PHP-CS-Fixer rules cannot be individually ignored. Imagine class with one readonly property. The class can be perfectly mutable class that is allowed to be extended by another non-readonly class.

**Comment by kubawerlos** [comment]
> Especially as PHP-CS-Fixer rules cannot be individually ignored

What do you mean? You can disable any rule you want.

**Comment by keradus** [comment]
I guess he mentioned "per-file".

**Comment by keradus** [comment]
I guess we can start with converting phpdoc tag to `readonly` keyword [ref https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/issues/6640 ]

**Comment by simensen** [comment]
> I do not think this is a good idea. Especially as PHP-CS-Fixer rules cannot be individually ignored. Imagine class with one readonly property. The class can be perfectly mutable class that is allowed to be extended by another non-readonly class.

If `readonly class` would be inherited by children, it would indeed cause potential issues. Could we make it configurable? For example, as I've learned more about `php-cs-fixer`, I'd imagine something like this could be part of the design? For example, default behavior would be to not offer to promote a class to readonly if it is abstract, but give the option to include those as well?

```php
[

  // Default behavior ( ['even_abstract' => false] )
  'promote_readonly_class' => true,

  // Can be overridden to 
  'promote_readonly_class' => [
    'even_abstract' => true,
  ],
```

I've spent very little time thinking about naming, here, but hopefully this helps show what I'm suggesting. :)

**Comment by Wirone** [comment]
Another option is to convert only `final` classes with all properties marked as `readonly`. For non-final classes we can break the code in inherited classes, so it's too risky IMHO.

**Comment by ByNetherdude** [comment]
+1 for the `final` option suggested by @Wirone.

**Comment by janisint** [comment]
still needed, please unstale

**Comment by hockdudu** [comment]
It appears there's a Rector rule for this: https://github.com/rectorphp/rector/blob/main/docs/rector_rules_overview.md#readonlyclassrector

**Comment by janisint** [comment]
Please keep this open

**Comment by Wirone** [comment]
@janisint there's no need to wait for green light from maintainers when they created and commented the issue, and [suggested how it should be implemented](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/issues/6801#issuecomment-1483813675) 🙂. Issue is open = accepted for implementation.

**Comment by gharlan** [comment]
> Another option is to convert only `final` classes with all properties marked as `readonly`. For non-final classes we can break the code in inherited classes, so it's too risky IMHO.

The classes handled by the fixer
* should be final
* should not extend other classes (cs-fixer does not know if parent class is readonly)
* should not use traits (cs-fixer does not know if trait contains mutable properties)
* should not contain static properties
* should not use `#[AllowDynamicProperties]` attribute (but even without the attribute it is risky, because using dynamic properties without the attribute is deprecated, but still possible)
